### ISSUE TITLE
Let actonc deduce system path based on source file and a new? acton system configuration file

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#39]
### Proposal
`actonc` should automatically detect the acton system path based on:
- the specified source file (the one we are asked to compile)
- from the specified source file, traverse directory structure upwards to find an acton configuration
- the acton config file is at the root of the system path, thus we have found system path

The configuration file can contain various project settings. I think this is sort of similar to how Rust's cargo works, e.g.:

```shell
$ cargo new my-project
     Created binary (application) `my-project` package
$ ls my-project
Cargo.toml
src
$ ls my-project/src
main.rs
```

when creating a new project, a top level project config file is generated (`Cargo.toml`) that contains project settings. Then there's a standardised project directory structure.

### Motivation
`actonc` currently assumes its own location is at the root of the system. This requires a copy of actonc with every Acton project. This in itself isn't overly bad, roughly similar to a Python virtualenv with its own interpreted (got to guarantee not just library version but also the interpreter!). However, like Python virtualens are an optional thing, I think it should be optional for Acton. I imagine that after you've installed acton, the compiler will live in /usr/bin along with its required supportive infrastructure in /usr/lib/acton or so. That should be enough to compile Acton programs.

@nordlander and I have talked about this or at least adjacent topics before and he convinced me about certain points - at least he had thought more about the topic than I had, at the time. I believe the core concept @nordlander wants is to avoid "path hell", where the compiled output will vary based on what paths happen to be in the hypothetical "ACTON_PATH". It seems to me that this boils down to not having a PATH for searching for libraries etc. Instead we want a package tool that resolves libraries for us, downloading specific version as necessary etc. I agree with this, but I think it is largely orthogonal to the location of `actonc`. We should not need to have a copy of actonc in itself. We can still consider `actonc` a package that should be versioned, or rather than in the project config, we can specify a particular version of `actonc` to use... but I don't think that needs to mean that it has to live in the root of project directory.

Haskell stack uses a specific resolver to resolve what versions of packages to use etc. The resolver also dictates the haskell compiler version. Stack downloads stuff to ~/.stack~ - so it is possible to have multiple versions installed, but they are not installed per project but rather in a shared location. This seems like a sane design. To allow it for acton, we need to decouple the installed location of acton with the project directory.

This is relevant right now because I want to create some form of "release" of Acton that can be downloaded by users wanting to try it out. The shape and form of that release will need to look different depending on how this works.

**Comment by plajjan** [comment]
It is hard talking about this without slipping into talking about the full package / module management for Acton, which I'm not sure we are ready for - at least not making any final decisions about. I figured we know enough to at least decide that `actonc` must be decoupled from the project directory.

**Comment by plajjan** [comment]
Blurb, looking at how to implement this, I realize the option is already there! We have a --path argument which is to specify the system path!

Still, this proposal is about automatically deriving it rather than having to specify it. Meanwhile I'm happy it is possible to specify, since that unblocks my work on building acton releases.

**Comment by plajjan** [comment]
And I also realize that I've misunderstood the meaning of "system". I figured it meant the resulting Acton system, like the outcome of a users Acton project, and thus it would contain the private modules we compile... but it naturally refers to the Acton system itself, like the compiler etc and the builtin / standard modules it comes with.

Spoke with @nordlander on this, what we agree on, where we disagree and how to move forward. I think I'll try to write a new issue to sum up the situation, with less misunderestandings this time ;)

**Comment by plajjan** [comment]
FWIW, in most places where I said "system" in here, I really mean "an Acton project"...

Anyway, closing this. Will rephrase!
